### PR TITLE
Feature/virtual terminal

### PR DESCRIPTION
**Pull request by andrii-onufriichuk**

### CODE REVIEW

File: Block/Order/View/Info.php [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    /**
+     * @return bool
+     * @throws LocalizedException
+     */
+    public function isMotoAvailable(): bool
+    {
+        $amount = $this->getOrder()->getGrandTotal();
+        $currency = $this->getOrder()->getOrderCurrencyCode();
+        if ($this->getOrder()->getPayment()->getMethod() == 'rvvup_virtual-terminal') {
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Block/Order/View/Info.php:105]: Yes, we can. 
Order must have payment method to be submitted

File: Block/Order/View/Info.php [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        return false;
+    }
+
+    /**
+     * @return bool
+     * @throws LocalizedException
+     */
+    public function isPaymentLinkOrder(): bool
+    {
+        if ($this->getOrder()->getPayment()->getMethod() == 'rvvup_payment-link') {
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Block/Order/View/Info.php:122]: Order must have payment method to be submitted

File: view/adminhtml/templates/order/view/info.phtml [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+                    <?php endif; ?>
+                </div>
+            </div>
+            <div class="admin__page-section-item-content">
+                <table class="admin__table-secondary order-account-information-table">
+                    <tr>
+                        <th><?= $escaper->escapeHtml(__('Customer Name')) ?></th>
+                        <td>
+                            <?php if ($customerUrl): ?>
+                                <a href="<?= $escaper->escapeUrl($customerUrl) ?>" target="_blank">
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply view/adminhtml/templates/order/view/info.phtml:136]: Those are not our modifications, it's copied from Magento default, search only for code inside 'Rvvup modifications' start/end

File: view/adminhtml/templates/order/view/info.phtml [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+                    <?php endif; ?>
+                    <?php if ($globalCurrencyCode !== $baseCurrencyCode): ?>
+                        <tr>
+                            <th><?= $escaper->escapeHtml(__('%1 / %2 rate:', $globalCurrencyCode, $baseCurrencyCode)) ?></th>
+                            <td><?= $escaper->escapeHtml($order->getBaseToGlobalRate()) ?></td>
+                        </tr>
+                    <?php endif; ?>
+                    <?php if ($baseCurrencyCode !== $orderCurrencyCode && $globalCurrencyCode !== $orderCurrencyCode): ?>
+                        <tr>
+                            <th><?= $escaper->escapeHtml(__('%1 / %2 rate:', $orderCurrencyCode, $baseCurrencyCode)) ?></th>
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply view/adminhtml/templates/order/view/info.phtml:110]: Those are not our modifications, it's copied from Magento default, search only for code inside 'Rvvup modifications' start/end

File: view/adminhtml/templates/order/view/info.phtml [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+                    <?php endif; ?>
+                    <?php if ($order->getRemoteIp() && $block->shouldDisplayCustomerIp()): ?>
+                        <tr>
+                            <th><?= $escaper->escapeHtml(__('Placed from IP')) ?></th>
+                            <td><?= $escaper->escapeHtml($order->getRemoteIp()); ?><?= $order->getXForwardedFor() ? ' (' . $escaper->escapeHtml($order->getXForwardedFor()) . ')' : ''; ?></td>
+                        </tr>
+                    <?php endif; ?>
+                    <?php if ($globalCurrencyCode !== $baseCurrencyCode): ?>
+                        <tr>
+                            <th><?= $escaper->escapeHtml(__('%1 / %2 rate:', $globalCurrencyCode, $baseCurrencyCode)) ?></th>
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply view/adminhtml/templates/order/view/info.phtml:104]: Those are not our modifications, it's copied from Magento default, search only for code inside 'Rvvup modifications' start/end

File: Controller/Adminhtml/Create/PaymentLink.php [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            ]);
+            return $result;
+        }
+
+        if ($amount > 0) {
+            try {
+                $order = $this->orderRepository->get($orderId);
+                $body = $this->paymentLinkService->createPaymentLink($storeId, $amount, $orderId, $currencyCode);
+                $message = $this->config->getPayByLinkText(ScopeInterface::SCOPE_STORE, $storeId);
+                $message .= PHP_EOL . $body['url'];
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Controller/Adminhtml/Create/PaymentLink.php:76]: PHP EOL is newline character, we add that to display link in the bottom which is easier to read and copy

File: Controller/Adminhtml/Create/PaymentLink.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                $this->logger->error('Failed to create Rvvup payment link from order_view page', [
+                    $amount,
+                    $storeId,
+                    $orderId,
+                    $currencyCode,
+                    $e->getMessage()
+                ]);
+
+                $result->setData([
+                    'success' => false
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Controller/Adminhtml/Create/PaymentLink.php:90]: If it fails, merchant will see an error and message to check the logs. 
It's not customer journey and the message likely to be huge sql statement so we put that to logs.

File: Controller/Adminhtml/Create/PaymentLink.php [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                'message' => 'Amount is empty!'
+            ]);
+            return $result;
+        }
+
+        $result->setData([
+            'success' => true
+        ]);
+
+        return $result;
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Controller/Adminhtml/Create/PaymentLink.php:106]: The indentation looks correct to me and PHPCS.
Adding another tab feels wrong as it's not nested to anything

File: Controller/Redirect/In.php [hunk 9]
```diff
@@ -79,10 +86,29 @@ public function execute()
     {
         $rvvupId = $this->request->getParam('rvvup-order-id');
         $paymentStatus = $this->request->getParam('payment-status');
-        $quote = $this->captureService->getQuoteByRvvupId($rvvupId);
+        $quote = $this->captureService->getQuoteByRvvupId((string) $rvvupId);
+        $checkoutId = $this->request->getParam('checkout_id');
+        $storeId = $this->request->getParam('store_id');
+
+        if ($checkoutId && $storeId) {
+            $order = $this->captureService->getOrderByPaymentField('rvvup_moto_id', $checkoutId, $storeId);
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Controller/Redirect/In.php]: Nice one! 
Pushed

File: Controller/Adminhtml/Create/VirtualTerminal.php [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            $result->setData([
+                'success' => false
+            ]);
+            return $result;
+        }
+
+        $result->setData([
+            'iframe-url' => $body['url'],
+            'success' => true
+        ]);
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Controller/Adminhtml/Create/VirtualTerminal.php]: Great spot!
Done

File: Controller/Redirect/In.php [hunk 11]
```diff
@@ -84,11 +84,12 @@ public function __construct(
      */
     public function execute()
     {
-        $rvvupId = $this->request->getParam(Method::ORDER_ID);
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Controller/Redirect/In.php]: Nope, that is two different ids (they do not match case)

File: view/adminhtml/templates/form/payment-link.phtml [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+$storeId = $quote->getStoreId();
+$currencyCode = $quote->getQuoteCurrencyCode();
+$url = $block->getUrl('rvvup/availability/index');
+?>
+<fieldset class="admin__fieldset payment-method"
+          id="payment_form_<?= $escaper->escapeHtmlAttr($code) ?>"
+          style="display:none;">
+    <div class="admin__field _required">
+        <label class="label admin__field-label" style="font-weight: normal; display: contents; word-break: auto-phrase;">
+            Will create and share in order email
```
Comment (jay-rvvup) [inline view/adminhtml/templates/form/payment-link.phtml:23]: Small text suggestion:
"The created link will be shared in the order email"

File: view/adminhtml/templates/form/virtual-terminal.phtml [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+$storeId = $quote->getStoreId();
+$currencyCode = $quote->getQuoteCurrencyCode();
+$url = $block->getUrl('rvvup/availability/index');
+?>
+<fieldset class="admin__fieldset payment-method"
+          id="payment_form_<?= $escaper->escapeHtmlAttr($code) ?>"
+          style="display:none;">
+    <div class="admin__field _required">
+        <label class="label admin__field-label" style="font-weight: normal; display: contents; word-break: auto-phrase;">
+            Available after placing order
```
Comment (jay-rvvup) [inline view/adminhtml/templates/form/virtual-terminal.phtml:23]: Small text suggestion:
"The virtual terminal will be available after placing the order"

File: Service/VirtualCheckout.php [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+            $payment = $this->paymentLinkService->getQuotePaymentByOrder($order);
+            $payment->setAdditionalInformation(Method::MOTO_ID, $motoId);
+
+            if ($payment->getAdditionalInformation(Method::PAYMENT_LINK_ID)) {
+                $paymentLinkId = $payment->getAdditionalInformation(Method::PAYMENT_LINK_ID);
+            } elseif ($order->getPayment()->getAdditionalInformation(Method::PAYMENT_LINK_ID)) {
+                $paymentLinkId = $order->getPayment()->getAdditionalInformation(Method::PAYMENT_LINK_ID);
+            }
+            if (isset($paymentLinkId)) {
+                $this->paymentLinkService->cancelPaymentLink((string)$order->getStoreId(), $paymentLinkId);
```
Comment (jay-rvvup) [inline Service/VirtualCheckout.php:233]: Do we still need this now they are two separate magento payment methods?

**Review by jay-rvvup — APPROVED** [review]

**Pull request merged**